From: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
Date: Tue, 1 Aug 2017 22:18:41 +0200
Subject: [PATCH] WIP: Bring in baremetal support

- start by copying reclass/classes/cluster/virtual-mcp-ocata-ovs as
  classes/cluster/baremetal-mcp-ocata-ovs;
- add new state (maas) that will handle MaaS configuration;

v1 -> v4:
- Split PXE network in two for baremetal:
  * rename old "pxe" virtual network to "mcpcontrol", make it
    non-configurable and identical for baremetal/virtual deploys;
  * new "pxebr" bridge is dedicated for MaaS fabric network, which
    comes with its own DHCP, TFTP etc.;
- Drop hardcoded PXE gateway & static IP for MaaS node, since
  "mcpcontrol" remains a NAT-ed virtual network, with its own DHCP;
- Keep internet access available on first interfaces for cfg01/mas01;
- Align MaaS IP addrs (all x.y.z.200), add public IP for easy debug
  via MaaS dashboard;
- Add static IP in new network segment (192.168.11.200/24) on MaaS
  node's PXE interface;
- Set MaaS PXE interface MTU 1500 (weird network errors with jumbo);
- sync_all is now upstream, drop from maas state;

TODO:
- salt_contorl_xenial_image is still harcoded to x86_64 image from
  apt-mk.mirantis.com;
- better seperation between cfg01 and mas01 configuration;
- make MaaS deployed nodes available to Salt master (either via
  statically configured mgmt IPs or routing between DHCP nets);

Change-Id: Ie8dd584b140991d2bd992acdfe47f5644bf51409
Signed-off-by: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
---
 ci/deploy.sh                                       |  11 +-
 mcp/config/scenario/baremetal/defaults.yaml        |   6 +
 .../baremetal/os-nosdn-nofeature-noha.yaml         |  14 ++
 mcp/config/states/maas                             |   9 ++
 .../haproxy_openstack_api.yml                      | 166 +++++++++++++++++++++
 .../cluster/baremetal-mcp-ocata-ovs/.gitkeep       |   0
 .../baremetal-mcp-ocata-ovs/infra/config.yml       |  77 ++++++++++
 .../cluster/baremetal-mcp-ocata-ovs/infra/init.yml |  19 +++
 .../cluster/baremetal-mcp-ocata-ovs/infra/maas.yml |  46 ++++++
 .../cluster/baremetal-mcp-ocata-ovs/init.yml       |  24 +++
 .../baremetal-mcp-ocata-ovs/openstack/compute.yml  |  81 ++++++++++
 .../baremetal-mcp-ocata-ovs/openstack/control.yml  |  98 ++++++++++++
 .../baremetal-mcp-ocata-ovs/openstack/gateway.yml  |  81 ++++++++++
 .../baremetal-mcp-ocata-ovs/openstack/init.yml     | 123 +++++++++++++++
 .../nodes/cfg01.baremetal-mcp-ocata-ovs.local.yml  |  10 ++
 mcp/scripts/lib.sh                                 |  50 ++++---
 mcp/scripts/net_mcpcontrol.xml                     |  10 ++
 mcp/scripts/net_pxe.xml                            |  10 --
 18 files changed, 800 insertions(+), 35 deletions(-)
 create mode 100644 mcp/config/scenario/baremetal/defaults.yaml
 create mode 100644 mcp/config/scenario/baremetal/os-nosdn-nofeature-noha.yaml
 create mode 100755 mcp/config/states/maas
 create mode 100644 mcp/reclass/classes/cluster/baremetal-mcp-ocata-common/haproxy_openstack_api.yml
 create mode 100644 mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/.gitkeep
 create mode 100644 mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/config.yml
 create mode 100644 mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/init.yml
 create mode 100644 mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/maas.yml
 create mode 100644 mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/init.yml
 create mode 100644 mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/compute.yml
 create mode 100644 mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/control.yml
 create mode 100644 mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/gateway.yml
 create mode 100644 mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/init.yml
 create mode 100644 mcp/reclass/nodes/cfg01.baremetal-mcp-ocata-ovs.local.yml
 create mode 100644 mcp/scripts/net_mcpcontrol.xml
 delete mode 100644 mcp/scripts/net_pxe.xml

diff --git a/ci/deploy.sh b/ci/deploy.sh
index 491b955..8893c3f 100755
--- a/ci/deploy.sh
+++ b/ci/deploy.sh
@@ -68,7 +68,11 @@ $(notify "Input parameters to the build script are:" 2)
    For an empty value, the deploy script will use virsh to create the default
    expected network (e.g. -B pxe,,,public will use existing "pxe" and "public"
    bridges, respectively create "mgmt" and "internal").
-   The default is pxebr.
+   Note that a virtual network "mcpcontrol" is always created. For virtual
+   deploys, "mcpcontrol" is also used for PXE, leaving the PXE bridge unused.
+   For baremetal deploys, PXE bridge is used for baremetal node provisioning,
+   while "mcpcontrol" is used to provision the infrastructure VMs only.
+   The default is 'pxebr'.
 -h Print this message and exit
 -l Lab name as defined in the configuration directory, e.g. lf
 -p POD name as defined in the configuration directory, e.g. pod-1
@@ -130,11 +134,12 @@ clean() {
 SCRIPT_PATH=$(readlink -f "$(dirname "${BASH_SOURCE[0]}")")
 DEPLOY_DIR=$(cd "${SCRIPT_PATH}/../mcp/scripts"; pwd)
 DEPLOY_TYPE='baremetal'
-OPNFV_BRIDGES=('pxe' 'mgmt' 'internal' 'public')
+OPNFV_BRIDGES=('pxebr' 'mgmt' 'internal' 'public')
 URI_REGEXP='(file|https?|ftp)://.*'

 export SSH_KEY=${SSH_KEY:-mcp.rsa}
 export SALT_MASTER=${SALT_MASTER_IP:-192.168.10.100}
+export MAAS_IP=${MAAS_IP:-192.168.10.200}
 export SSH_OPTS="-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -i ${SSH_KEY}"

 # Variables below are disabled for now, to be re-introduced or removed later
@@ -311,7 +316,7 @@ generate_ssh_key
 prepare_vms virtual_nodes "${base_image}"
 create_networks OPNFV_BRIDGES
 create_vms virtual_nodes virtual_nodes_ram virtual_nodes_vcpus OPNFV_BRIDGES
-update_pxe_network OPNFV_BRIDGES
+update_mcpcontrol_network
 start_vms virtual_nodes
 check_connection

diff --git a/mcp/config/scenario/baremetal/defaults.yaml b/mcp/config/scenario/baremetal/defaults.yaml
new file mode 100644
index 0000000..b841e88
--- /dev/null
+++ b/mcp/config/scenario/baremetal/defaults.yaml
@@ -0,0 +1,6 @@
+base_image: https://cloud-images.ubuntu.com/xenial/current/xenial-server-cloudimg-amd64-disk1.img
+virtual:
+  default:
+    vcpus: 2
+    ram: 4096
+
diff --git a/mcp/config/scenario/baremetal/os-nosdn-nofeature-noha.yaml b/mcp/config/scenario/baremetal/os-nosdn-nofeature-noha.yaml
new file mode 100644
index 0000000..10ad159
--- /dev/null
+++ b/mcp/config/scenario/baremetal/os-nosdn-nofeature-noha.yaml
@@ -0,0 +1,14 @@
+cluster:
+  domain: baremetal-mcp-ocata-ovs.local
+  states:
+    - maas
+    - openstack
+    - neutron_compute
+    - networks
+virtual:
+  nodes:
+    - cfg01
+    - mas01
+  mas01:
+    vcpus: 4
+    ram: 16384
diff --git a/mcp/config/states/maas b/mcp/config/states/maas
new file mode 100755
index 0000000..bed3823
--- /dev/null
+++ b/mcp/config/states/maas
@@ -0,0 +1,9 @@
+salt 'mas01*' cmd.run "add-apt-repository ppa:maas/stable"
+
+salt 'mas01*' state.apply linux,salt,openssh,ntp
+salt 'mas01*' state.apply linux.network.interface
+salt 'mas01*' state.apply maas.cluster,maas.region
+
+salt 'mas01*' pillar.item\
+  maas:region:admin:username \
+  maas:region:admin:password
diff --git a/mcp/reclass/classes/cluster/baremetal-mcp-ocata-common/haproxy_openstack_api.yml b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-common/haproxy_openstack_api.yml
new file mode 100644
index 0000000..e63e9d5
--- /dev/null
+++ b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-common/haproxy_openstack_api.yml
@@ -0,0 +1,166 @@
+parameters:
+  _param:
+    haproxy_check: check inter 15s fastinter 2s downinter 4s rise 3 fall 3
+  haproxy:
+    proxy:
+      listen:
+        cinder_api:
+          type: openstack-service
+          service_name: cinder
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 8776
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 8776
+            params: ${_param:haproxy_check}
+        glance_api:
+          type: openstack-service
+          service_name: glance
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 9292
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 9292
+            params: ${_param:haproxy_check}
+        glance_registry_api:
+          type: general-service
+          service_name: glance
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 9191
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 9191
+            params: ${_param:haproxy_check}
+        glare:
+          type: general-service
+          service_name: glare
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 9494
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 9494
+            params: ${_param:haproxy_check}
+        heat_cloudwatch_api:
+          type: openstack-service
+          service_name: heat
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 8003
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 8003
+            params: ${_param:haproxy_check}
+        heat_api:
+          type: openstack-service
+          service_name: heat
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 8004
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 8004
+            params: ${_param:haproxy_check}
+        heat_cfn_api:
+          type: openstack-service
+          service_name: heat
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 8000
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 8000
+            params: ${_param:haproxy_check}
+        keystone_public_api:
+          type: openstack-service
+          service_name: keystone
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 5000
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 5000
+            params: ${_param:haproxy_check}
+        keystone_admin_api:
+          type: openstack-service
+          service_name: keystone
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 35357
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 35357
+            params: ${_param:haproxy_check}
+        neutron_api:
+          type: openstack-service
+          service_name: neutron
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 9696
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 9696
+            params: ${_param:haproxy_check}
+        nova_placement_api:
+          mode: http
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 8778
+          options:
+          - httpclose
+          - httplog
+          health-check:
+            http:
+              options:
+              - expect status 401
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 8778
+            params: ${_param:haproxy_check}
+        nova_ec2_api:
+          type: general-service
+          service_name: nova
+          check: false
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 8773
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 8773
+            params: ${_param:haproxy_check}
+        nova_api:
+          type: openstack-service
+          service_name: nova
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 8774
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 8774
+            params: ${_param:haproxy_check}
+        nova_metadata_api:
+          type: openstack-service
+          binds:
+          - address: ${_param:cluster_vip_address}
+            port: 8775
+          servers:
+          - name: ctl01
+            host: ${_param:cluster_node01_address}
+            port: 8775
+            params: ${_param:haproxy_check}
diff --git a/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/.gitkeep b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/.gitkeep
new file mode 100644
index 0000000..e69de29
diff --git a/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/config.yml b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/config.yml
new file mode 100644
index 0000000..a1b021a
--- /dev/null
+++ b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/config.yml
@@ -0,0 +1,77 @@
+classes:
+- service.git.client
+- system.linux.system.single
+- system.linux.system.repo.mcp.salt
+- system.salt.master.api
+- system.salt.master.pkg
+- system.reclass.storage.salt
+- system.salt.minion.ca.salt_master
+- system.salt.minion.cert.proxy
+- system.mysql.client.single
+- system.reclass.storage.system.openstack_compute_multi
+- system.reclass.storage.system.openstack_gateway_single
+- system.reclass.storage.system.infra_maas_single
+- system.salt.control.cluster.infra_maas_single
+- cluster.baremetal-mcp-ocata-ovs
+parameters:
+  _param:
+    infra_maas_node01_hostname: mas01
+    infra_kvm_node01_hostname: kvm01
+    salt_control_xenial_image: http://apt-mk.mirantis.com/images/ubuntu-16-04-x64-mcp1.1.qcow2
+    openstack_control_node01_hostname: ctl01
+    reclass_data_repository: local
+    salt_master_environment_repository: "https://github.com/tcpcloud"
+    salt_master_environment_revision: master
+    reclass_config_master: 192.168.10.100
+    single_address: 172.16.10.100
+    salt_master_host: 127.0.0.1
+    salt_master_base_environment: prd
+    salt_minion_ca_host: ${linux:network:fqdn}
+    salt_api_password_hash: "$6$sGnRlxGf$al5jMCetLP.vfI/fTl3Z0N7Za1aeiexL487jAtyRABVfT3NlwZxQGVhO7S1N8OwS/34VHYwZQA8lkXwKMN/GS1"
+  linux:
+    network:
+      interface:
+        ens4:
+          enabled: true
+          type: eth
+          proto: static
+          address: ${_param:single_address}
+          netmask: 255.255.255.0
+  reclass:
+    storage:
+      data_source:
+        engine: local
+      node:
+        infra_maas_node01:
+          name: ${_param:infra_maas_node01_hostname}
+          domain: ${_param:cluster_domain}
+          classes:
+          - cluster.${_param:cluster_name}.infra.maas
+          params:
+            salt_master_host: ${_param:reclass_config_master}
+            single_address: ${_param:infra_maas_node01_deploy_address}
+            pxe_address: 192.168.11.200
+            external_address: 10.16.0.200
+        openstack_control_node01:
+          name: ${_param:openstack_control_node01_hostname}
+          domain: ${_param:cluster_domain}
+          classes:
+          - cluster.${_param:cluster_name}.openstack.control
+          params:
+            linux_system_codename: xenial
+            salt_master_host: ${_param:reclass_config_master}
+            single_address: ${_param:openstack_control_node01_address}
+        openstack_compute_node01:
+          params:
+            single_address: 172.16.10.105
+            tenant_address: 10.1.0.105
+            external_address: 10.16.0.105
+        openstack_compute_node02:
+          params:
+            single_address: 172.16.10.106
+            tenant_address: 10.1.0.106
+            external_address: 10.16.0.106
+        openstack_gateway_node01:
+          params:
+            tenant_address: 10.1.0.110
+            external_address: 10.16.0.110
diff --git a/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/init.yml b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/init.yml
new file mode 100644
index 0000000..90fc9cf
--- /dev/null
+++ b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/init.yml
@@ -0,0 +1,19 @@
+parameters:
+  linux:
+    network:
+      host:
+        cfg01:
+          address: ${_param:infra_config_address}
+          names:
+          - cfg01
+          - cfg01.${_param:cluster_domain}
+        cfg:
+          address: ${_param:infra_config_address}
+          names:
+          - cfg
+          - cfg.${_param:cluster_domain}
+        mas01:
+          address: ${_param:infra_maas_node01_deploy_address}
+          names:
+          - mas01
+          - mas01.${_param:cluster_domain}
diff --git a/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/maas.yml b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/maas.yml
new file mode 100644
index 0000000..9f0d8bb
--- /dev/null
+++ b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/infra/maas.yml
@@ -0,0 +1,46 @@
+classes:
+- system.linux.system.repo.saltstack.xenial
+- system.maas.region.single
+- cluster.baremetal-mcp-ocata-ovs
+parameters:
+  _param:
+    primary_interface: ens4
+    pxe_interface: ens5
+    external_interface: ens6
+    interface_mtu: 9000
+    pxe_interface_mtu: 1500
+    linux_system_codename: xenial
+    maas_admin_username: opnfv
+    maas_admin_password: opnfv_secret
+    maas_db_password: opnfv_secret
+    dns_server01: 8.8.8.8
+  maas:
+    region:
+      salt_master_ip: 192.168.10.100
+  linux:
+    network:
+      interface:
+        primary_interface:
+          enabled: true
+          name: ${_param:primary_interface}
+          mtu: ${_param:interface_mtu}
+          proto: static
+          address: ${_param:single_address}
+          netmask: 255.255.255.0
+          type: eth
+        pxe_interface:
+          enabled: true
+          name: ${_param:pxe_interface}
+          mtu: ${_param:pxe_interface_mtu}
+          proto: static
+          address: ${_param:pxe_address}
+          netmask: 255.255.255.0
+          type: eth
+        external_interface:
+          enabled: true
+          name: ${_param:external_interface}
+          mtu: ${_param:interface_mtu}
+          proto: static
+          address: ${_param:external_address}
+          netmask: 255.255.255.0
+          type: eth
diff --git a/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/init.yml b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/init.yml
new file mode 100644
index 0000000..a539038
--- /dev/null
+++ b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/init.yml
@@ -0,0 +1,24 @@
+classes:
+- system.linux.system.single
+- cluster.baremetal-mcp-ocata-ovs.infra
+- cluster.baremetal-mcp-ocata-ovs.openstack
+
+parameters:
+  _param:
+    cluster_domain: baremetal-mcp-ocata-ovs.local
+    cluster_name: baremetal-mcp-ocata-ovs
+    # infra service addresses
+    infra_config_address: 172.16.10.100
+    infra_maas_node01_deploy_address: 172.16.10.200
+    # openstack service addresses
+    openstack_control_address: 172.16.10.101
+    openstack_control_node01_address: 172.16.10.101
+    openstack_control_node02_address: 172.16.10.102
+    openstack_control_node03_address: 172.16.10.103
+    openstack_database_address: ${_param:openstack_control_address}
+    openstack_message_queue_address: ${_param:openstack_control_address}
+    openstack_message_queue_node01_address: ${_param:openstack_control_node01_address}
+    openstack_message_queue_node02_address: ${_param:openstack_control_node02_address}
+    openstack_message_queue_node03_address: ${_param:openstack_control_node03_address}
+    openstack_gateway_address: 172.16.10.110
+    control_address: ${_param:openstack_control_address}
diff --git a/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/compute.yml b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/compute.yml
new file mode 100644
index 0000000..d9c2db7
--- /dev/null
+++ b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/compute.yml
@@ -0,0 +1,81 @@
+classes:
+- system.linux.system.repo.mcp.openstack
+- system.linux.system.repo.mcp.extra
+- system.linux.system.repo.saltstack.xenial
+- system.linux.storage.loopback
+- system.nova.compute.single
+- service.neutron.compute.single
+- service.cinder.volume.single
+- system.cinder.volume.backend.lvm
+- cluster.baremetal-mcp-ocata-ovs
+parameters:
+  _param:
+    primary_interface: ens4
+    tenant_interface: ens5
+    external_interface: ens6
+    interface_mtu: 9000
+    linux_system_codename: xenial
+    loopback_device_size: 10
+  nova:
+    compute:
+      vncproxy_url: http://${_param:cluster_vip_address}:6080
+      network:
+        region: ${_param:openstack_region}
+        user: neutron
+        tenant: service
+        password: ${_param:keystone_neutron_password}
+  neutron:
+    compute:
+      agent_mode: ${_param:neutron_compute_agent_mode}
+      message_queue:
+        host: ${_param:openstack_control_address}
+      metadata:
+        host: ${_param:openstack_control_address}
+  cinder:
+    volume:
+      database:
+        host: ${_param:cluster_local_address}
+      identity:
+        host: ${_param:cluster_local_address}
+      glance:
+        host: ${_param:cluster_local_address}
+      message_queue:
+        host: ${_param:cluster_local_address}
+  linux:
+    network:
+      bridge: openvswitch
+      interface:
+        dhcp_int:
+          enabled: true
+          name: ens3
+          proto: dhcp
+          type: eth
+          mtu: ${_param:interface_mtu}
+        primary_interface:
+          enabled: true
+          name: ${_param:primary_interface}
+          mtu: ${_param:interface_mtu}
+          proto: manual
+          type: eth
+        tenant_interface:
+          enabled: true
+          name: ${_param:tenant_interface}
+          mtu: ${_param:interface_mtu}
+          proto: manual
+          type: eth
+        br-mgmt:
+          enabled: true
+          type: bridge
+          proto: static
+          address: ${_param:single_address}
+          netmask: 255.255.255.0
+          use_interfaces:
+          - ${_param:primary_interface}
+        br-mesh:
+          enabled: true
+          type: bridge
+          proto: static
+          address: ${_param:tenant_address}
+          netmask: 255.255.255.0
+          use_interfaces:
+          - ${_param:tenant_interface}
diff --git a/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/control.yml b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/control.yml
new file mode 100644
index 0000000..3bd7238
--- /dev/null
+++ b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/control.yml
@@ -0,0 +1,98 @@
+classes:
+- system.linux.system.lowmem
+- system.linux.system.repo.mcp.openstack
+- system.linux.system.repo.mcp.extra
+- system.linux.system.repo.saltstack.xenial
+- system.memcached.server.single
+- system.rabbitmq.server.single
+- system.rabbitmq.server.vhost.openstack
+- system.keystone.server.wsgi
+- system.keystone.server.single
+- system.keystone.client.single
+- system.keystone.client.service.nova21
+- system.keystone.client.service.nova-placement
+- system.keystone.client.service.glare
+- system.keystone.client.service.cinder3
+- system.glance.control.single
+- system.nova.control.single
+- system.neutron.control.openvswitch.single
+- system.cinder.control.single
+- system.cinder.control.backend.lvm
+- system.heat.server.single
+- service.mysql.server.single
+- system.galera.server.database.cinder
+- system.galera.server.database.glance
+- system.galera.server.database.grafana
+- system.galera.server.database.heat
+- system.galera.server.database.keystone
+- system.galera.server.database.nova
+- system.horizon.server.single
+- service.haproxy.proxy.single
+- cluster.baremetal-mcp-ocata-common.haproxy_openstack_api
+- cluster.baremetal-mcp-ocata-ovs
+parameters:
+  _param:
+    linux_system_codename: xenial
+  linux:
+    system:
+      package:
+        python-msgpack:
+          version: latest
+    network:
+      interface:
+        ens4:
+          enabled: true
+          type: eth
+          proto: static
+          address: ${_param:single_address}
+          netmask: 255.255.255.0
+        ens6:
+          enabled: true
+          type: eth
+          proto: static
+          address: ${_param:cluster_public_host}
+          netmask: 255.255.255.0
+  keystone:
+    server:
+      admin_email: ${_param:admin_email}
+      pkgs:
+      - keystone
+      - python-keystone
+      - python-keystoneclient
+      - python-psycopg2
+      - python-mysqldb
+      - python-six
+      - python-memcache
+      - python-openstackclient
+      - gettext-base
+      - python-pycadf
+  glance:
+    server:
+      storage:
+        engine: file
+      images: []
+      workers: 1
+  nova:
+    controller:
+      networking: dvr
+      cpu_allocation: 54
+      metadata:
+        password: ${_param:metadata_password}
+      bind:
+        private_address: ${_param:cluster_local_address}
+        public_address: ${_param:cluster_vip_address}
+        novncproxy_port: 6080
+      vncproxy_url: http://${_param:cluster_vip_address}:6080
+      workers: 1
+  heat:
+    server:
+      bind:
+        api_cfn:
+          address: ${_param:single_address}
+        api_cloudwatch:
+          address: ${_param:single_address}
+  mysql:
+    server:
+      version: '5.7'
+      bind:
+        address: ${_param:cluster_local_address}
diff --git a/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/gateway.yml b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/gateway.yml
new file mode 100644
index 0000000..38ca78c
--- /dev/null
+++ b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/gateway.yml
@@ -0,0 +1,81 @@
+classes:
+- system.linux.system.repo.mcp.openstack
+- system.linux.system.repo.mcp.extra
+- system.linux.system.repo.saltstack.xenial
+- service.neutron.gateway.single
+- cluster.baremetal-mcp-ocata-ovs
+parameters:
+  _param:
+    primary_interface: ens4
+    tenant_interface: ens5
+    external_interface: ens6
+    interface_mtu: 9000
+    linux_system_codename: xenial
+  neutron:
+    gateway:
+      agent_mode: ${_param:neutron_gateway_agent_mode}
+  linux:
+    network:
+      bridge: openvswitch
+      interface:
+        dhcp_int:
+          enabled: true
+          name: ens3
+          proto: dhcp
+          type: eth
+          mtu: ${_param:interface_mtu}
+        primary_interface:
+          enabled: true
+          name: ${_param:primary_interface}
+          mtu: ${_param:interface_mtu}
+          proto: manual
+          type: eth
+        tenant_interface:
+          enabled: true
+          name: ${_param:tenant_interface}
+          mtu: ${_param:interface_mtu}
+          proto: manual
+          type: eth
+        external_interface:
+          enabled: true
+          name: ${_param:external_interface}
+          mtu: ${_param:interface_mtu}
+          proto: manual
+          type: eth
+        br-floating:
+          enabled: true
+          type: ovs_bridge
+          mtu: ${_param:interface_mtu}
+        br-mgmt:
+          enabled: true
+          type: bridge
+          proto: static
+          address: ${_param:single_address}
+          netmask: 255.255.255.0
+          mtu: ${_param:interface_mtu}
+          use_interfaces:
+          - ${_param:primary_interface}
+        br-mesh:
+          enabled: true
+          type: bridge
+          mtu: ${_param:interface_mtu}
+          proto: static
+          address: ${_param:tenant_address}
+          netmask: 255.255.255.0
+          use_interfaces:
+          - ${_param:tenant_interface}
+        float-to-ex:
+          enabled: true
+          type: ovs_port
+          mtu: ${_param:interface_mtu}
+          bridge: br-floating
+        br-ex:
+          enabled: true
+          type: bridge
+          mtu: ${_param:interface_mtu}
+          address: ${_param:external_address}
+          netmask: 255.255.255.0
+          use_interfaces:
+          - ${_param:external_interface}
+          use_ovs_ports:
+          - float-to-ex
diff --git a/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/init.yml b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/init.yml
new file mode 100644
index 0000000..7079fd1
--- /dev/null
+++ b/mcp/reclass/classes/cluster/baremetal-mcp-ocata-ovs/openstack/init.yml
@@ -0,0 +1,123 @@
+parameters:
+  _param:
+    openstack_version: ocata
+    apt_mk_version: nightly
+    mcp_repo_version: 1.1
+    openstack_region: RegionOne
+    admin_email: root@localhost
+    cluster_public_protocol: http
+    cluster_public_host: 10.16.0.101
+    neutron_public_protocol: http
+    neutron_control_dvr: False
+    neutron_tenant_network_types: "flat,vxlan"
+    neutron_l3_ha: False
+    neutron_global_physnet_mtu: 1500
+    neutron_external_mtu: 1500
+    neutron_gateway_dvr: False
+    neutron_gateway_agent_mode: legacy
+    neutron_compute_dvr: False
+    neutron_compute_agent_mode: legacy
+    neutron_compute_external_access: False
+    galera_server_cluster_name: openstack_cluster
+    galera_server_maintenance_password: opnfv_secret
+    galera_server_admin_password: opnfv_secret
+    cluster_vip_address: ${_param:cluster_public_host}
+    cluster_local_address: ${_param:openstack_control_address}
+    cluster_node01_hostname: ctl01
+    cluster_node01_address: 172.16.10.101
+    cluster_node02_hostname: ctl02
+    cluster_node02_address: 172.16.10.102
+    cluster_node03_hostname: ctl03
+    cluster_node03_address: 172.16.10.103
+    rabbitmq_secret_key: opnfv_secret
+    rabbitmq_admin_password: opnfv_secret
+    rabbitmq_openstack_password: opnfv_secret
+    rabbitmq_cold_password: opnfv_secret
+    glance_version: ${_param:openstack_version}
+    glance_service_host: ${_param:cluster_local_address}
+    keystone_version: ${_param:openstack_version}
+    keystone_service_host: ${_param:cluster_local_address}
+    heat_version: ${_param:openstack_version}
+    heat_service_host: ${_param:cluster_local_address}
+    heat_domain_admin_password: opnfv_secret
+    ceilometer_version: ${_param:openstack_version}
+    ceilometer_service_host: 172.16.10.108
+    ceilometer_database_host: ${_param:cluster_local_address}
+    cinder_version: ${_param:openstack_version}
+    cinder_service_host: ${_param:cluster_local_address}
+    ceilometer_graphite_publisher_host: 172.16.10.107
+    ceilometer_graphite_publisher_port: 2013
+    nova_version: ${_param:openstack_version}
+    nova_service_host: ${_param:cluster_local_address}
+    nova_vncproxy_url: http://${_param:cluster_vip_address}:8060
+    neutron_version: ${_param:openstack_version}
+    neutron_service_host: ${_param:cluster_local_address}
+    metadata_password: password
+    mysql_admin_user: root
+    mysql_admin_password: opnfv_secret
+    mysql_cinder_password: opnfv_secret
+    mysql_ceilometer_password: opnfv_secret
+    mysql_glance_password: opnfv_secret
+    mysql_grafana_password: opnfv_secret
+    mysql_heat_password: opnfv_secret
+    mysql_keystone_password: opnfv_secret
+    mysql_neutron_password: opnfv_secret
+    mysql_nova_password: opnfv_secret
+    mysql_aodh_password: opnfv_secret
+    keystone_service_token: opnfv_secret
+    keystone_admin_password: opnfv_secret
+    keystone_ceilometer_password: opnfv_secret
+    keystone_cinder_password: opnfv_secret
+    keystone_glance_password: opnfv_secret
+    keystone_heat_password: opnfv_secret
+    keystone_keystone_password: opnfv_secret
+    keystone_neutron_password: opnfv_secret
+    keystone_nova_password: opnfv_secret
+    ceilometer_secret_key: opnfv_secret
+    metadata_password: opnfv_secret
+    horizon_version: ${_param:openstack_version}
+    horizon_secret_key: opaesee8Que2yahJoh9fo0eefo1Aeyo6ahyei8zeiboh3aeth5loth7ieNa5xi5e
+    horizon_identity_host: ${_param:cluster_vip_address}
+    horizon_identity_encryption: none
+    horizon_identity_version: 3
+    mongodb_server_replica_set: ceilometer
+    mongodb_ceilometer_password: cloudlab
+    mongodb_admin_password: cloudlab
+    mongodb_shared_key: eoTh1AwahlahqueingeejooLughah4tei9feing0eeVaephooDi2li1TaeV1ooth
+    aodh_version: ${_param:openstack_version}
+    keystone_aodh_password: opnfv_secret
+    aodh_service_host: 172.16.10.108
+  linux:
+    system:
+      kernel:
+        sysctl:
+          net.ipv4.tcp_congestion_control: yeah
+          net.ipv4.tcp_slow_start_after_idle: 0
+          net.ipv4.tcp_fin_timeout: 30
+    network:
+      host:
+        ctl:
+          address: ${_param:openstack_control_address}
+          names:
+          - ctl
+          - ctl.${_param:cluster_domain}
+        ctl01:
+          address: ${_param:openstack_control_node01_address}
+          names:
+          - ctl01
+          - ctl01.${_param:cluster_domain}
+        gtw01:
+          address: ${_param:openstack_gateway_address}
+          names:
+          - gtw01
+          - gtw01.${_param:cluster_domain}
+        cmp01:
+          address: 172.16.10.105
+          names:
+          - cmp01
+          - cmp01.${_param:cluster_domain}
+        cmp02:
+          address: 172.16.10.106
+          names:
+          - cmp02
+          - cmp02.${_param:cluster_domain}
diff --git a/mcp/reclass/nodes/cfg01.baremetal-mcp-ocata-ovs.local.yml b/mcp/reclass/nodes/cfg01.baremetal-mcp-ocata-ovs.local.yml
new file mode 100644
index 0000000..fd8b0a0
--- /dev/null
+++ b/mcp/reclass/nodes/cfg01.baremetal-mcp-ocata-ovs.local.yml
@@ -0,0 +1,10 @@
+classes:
+- cluster.baremetal-mcp-ocata-ovs.infra.config
+parameters:
+  _param:
+    linux_system_codename: xenial
+    reclass_data_revision: master
+  linux:
+    system:
+      name: cfg01
+      domain: baremetal-mcp-ocata-ovs.local
diff --git a/mcp/scripts/lib.sh b/mcp/scripts/lib.sh
index 8d45100..28b11e1 100644
--- a/mcp/scripts/lib.sh
+++ b/mcp/scripts/lib.sh
@@ -46,10 +46,12 @@ prepare_vms() {

 create_networks() {
   local -n vnode_networks=$1
-  # create required networks
-  for net in "${vnode_networks[@]}"; do
+  # create required networks, including constant "mcpcontrol"
+  # FIXME(alav): since we renamed "pxe" to "mcpcontrol", we need to make sure
+  # we delete the old "pxe" virtual network, or it would cause IP conflicts.
+  for net in "pxe" "mcpcontrol" "${vnode_networks[@]}"; do
     if virsh net-info "${net}" >/dev/null 2>&1; then
-      virsh net-destroy "${net}"
+      virsh net-destroy "${net}" || true
       virsh net-undefine "${net}"
     fi
     # in case of custom network, host should already have the bridge in place
@@ -67,17 +69,6 @@ create_vms() {
   local -n vnodes_vcpus=$3
   local -n vnode_networks=$4

-  # prepare network args
-  net_args=""
-  for net in "${vnode_networks[@]}"; do
-    net_type="network"
-    # in case of custom network, host should already have the bridge in place
-    if [ ! -f "net_${net}.xml" ]; then
-      net_type="bridge"
-    fi
-    net_args="${net_args} --network ${net_type}=${net},model=virtio"
-  done
-
   # AArch64: prepare arch specific arguments
   local virt_extra_args=""
   if [ "$(uname -i)" = "aarch64" ]; then
@@ -87,6 +78,21 @@ create_vms() {

   # create vms with specified options
   for node in "${vnodes[@]}"; do
+    # prepare network args
+    net_args=" --network network=mcpcontrol,model=virtio"
+    if [ "${node}" = "mas01" ]; then
+      # MaaS node's 3rd interface gets connected to PXE/Admin Bridge
+      vnode_networks[2]="${vnode_networks[0]}"
+    fi
+    for net in "${vnode_networks[@]:1}"; do
+      net_type="network"
+      # in case of custom network, host should already have the bridge in place
+      if [ ! -f "net_${net}.xml" ]; then
+        net_type="bridge"
+      fi
+      net_args="${net_args} --network ${net_type}=${net},model=virtio"
+    done
+
     # shellcheck disable=SC2086
     virt-install --name "${node}" \
     --ram "${vnodes_ram[$node]}" --vcpus "${vnodes_vcpus[$node]}" \
@@ -100,14 +106,14 @@ create_vms() {
   done
 }

-update_pxe_network() {
-  local -n vnode_networks=$1
-  if virsh net-info "${vnode_networks[0]}" >/dev/null 2>&1; then
-    # set static ip address for salt master node, only if managed via virsh
-    # NOTE: below expr assume PXE network is always the first in domiflist
-    virsh net-update "${vnode_networks[0]}" add ip-dhcp-host \
-    "<host mac='$(virsh domiflist cfg01 | awk '/network/ {print $5; exit}')' name='cfg01' ip='${SALT_MASTER}'/>" --live
-  fi
+update_mcpcontrol_network() {
+  # set static ip address for salt master node, MaaS node
+  local cmac=$(virsh domiflist cfg01 2>&1| awk '/mcpcontrol/ {print $5; exit}')
+  local amac=$(virsh domiflist mas01 2>&1| awk '/mcpcontrol/ {print $5; exit}')
+  virsh net-update "mcpcontrol" add ip-dhcp-host \
+    "<host mac='${cmac}' name='cfg01' ip='${SALT_MASTER}'/>" --live
+  [ -z "${amac}" ] || virsh net-update "mcpcontrol" add ip-dhcp-host \
+    "<host mac='${amac}' name='mas01' ip='${MAAS_IP}'/>" --live
 }

 start_vms() {
diff --git a/mcp/scripts/net_mcpcontrol.xml b/mcp/scripts/net_mcpcontrol.xml
new file mode 100644
index 0000000..f756ee0
--- /dev/null
+++ b/mcp/scripts/net_mcpcontrol.xml
@@ -0,0 +1,10 @@
+<network>
+  <name>mcpcontrol</name>
+  <bridge name="mcpcontrol"/>
+  <forward mode="nat"/>
+  <ip address="192.168.10.1" netmask="255.255.255.0">
+    <dhcp>
+      <range start="192.168.10.100" end="192.168.10.254"/>
+    </dhcp>
+  </ip>
+</network>
diff --git a/mcp/scripts/net_pxe.xml b/mcp/scripts/net_pxe.xml
deleted file mode 100644
index 92eaa6b..0000000
--- a/mcp/scripts/net_pxe.xml
+++ /dev/null
@@ -1,10 +0,0 @@
-<network>
-  <name>pxe</name>
-  <bridge name="pxe"/>
-  <forward mode="nat"/>
-  <ip address="192.168.10.1" netmask="255.255.255.0">
-    <dhcp>
-      <range start="192.168.10.100" end="192.168.10.254"/>
-    </dhcp>
-  </ip>
-</network>
